### PR TITLE
New llm as a judge notebook

### PR DESCRIPTION
**Pull request by joaq-rod**

### CODE REVIEW

File: examples/llm-as-a-judge.ipynb [hunk 1]
```diff
@@ -0,0 +1,676 @@
+{
```
Comment (rymc) [inline examples/llm-as-a-judge.ipynb:1]: <blockquote>timeconsuming</blockquote><p>time-consuming</p><p> </p><blockquote>Even more if our model is predicting large volumes of data.</blockquote><p>We can probably drop this sentence given the previous.</p><p> </p><blockquote>&nbsp;and provide an evaluating base to understand how the base model is performing.</blockquote><p>I find this sentence hard to follow. The concept of base model is a little overloaded in LLMs (base vs instruct), but I understand it's also used differently in LLM as a judge. Perhaps we can explicitly say that we will use model X as the "base model" producing predictions, and model Y as the "judge model", i.e., making it clear to readers what base model means and how we're using them. </p><p> </p><blockquote>&nbsp;As in this case we chose the IMDb 1000 Movie dataset which actually provides the ground truth we'll be able to check how good is our judge compared with the original ground truth that comes with the dataset.</blockquote><p>This sentence is a little awkward to read.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kluster-ai/klusterai-cookbook/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (rymc) [inline examples/llm-as-a-judge.ipynb:1]: <p>I think this would really benefit from a small amount of text introducing what is happening in each step/cell.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kluster-ai/klusterai-cookbook/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (rymc) [inline examples/llm-as-a-judge.ipynb:1]: <p>I think this would benefit from some text introducing what is happening in the steps. We can also link to docs / other notebooks too -- but it would be helpful to have a concise overview of what is happening in each step.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kluster-ai/klusterai-cookbook/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (rymc) [inline examples/llm-as-a-judge.ipynb:1]: <blockquote>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;"response_format": {"type": "json_object"},</blockquote><p>we should remove this from all notebooks as it is currently not supported.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kluster-ai/klusterai-cookbook/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (rymc) [inline examples/llm-as-a-judge.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    def create_tasks(df, task_type, system_prompt, model):
</pre><p>would it be clearer if we used "content" as an argument? i think it would be easier to follow and more reusable too!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kluster-ai/klusterai-cookbook/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (rymc) [inline examples/llm-as-a-judge.ipynb:1]: <p>I think this notebook would benefit from using two different "base models" to compare performance. It would then highlight the usefulness of the LLM as a judge to help make a decision on what model to use? </p><p> </p><p>Alternatively, trying some different prompts could work here instead. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/kluster-ai/klusterai-cookbook/pull/17/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

**Comment by rymc** [comment]
beyond the comments above, I think this notebook could be reviewed with some more text describing each step throughout, perhaps a more concise version of the descriptions in https://github.com/kluster-ai/klusterai-cookbook/blob/main/examples/model-comparison.ipynb

File: examples/llm-as-a-judge.ipynb (hunk 1 above)
Comment (joaq-rod) [reply examples/llm-as-a-judge.ipynb:1]: <p>How about this?</p><p>'''</p><p>In our previous &lt;a href="<a href="https://github.com/kluster-ai/klusterai-cookbook/blob/main/examples/model-comparison.ipynb%22" target="_blank">https://github.com/kluster-ai/klusterai-cookbook/blob/main/examples/model-comparison.ipynb"</a> target="_blank"&gt;notebook&lt;/a&gt; we explored the idea of selecting the best model to perform a classification task. We did that by calculating the accuracy of each model based on a ground truth label. In real life applications, though, the ground truth is not always available and to create one we might depend on human anotation which is time-consuming and costly.&nbsp;</p><p> </p><p>In this notebook, we will use the <code>Llama-3.1-8B-Instruct-Turbo</code> model to classify the genre of movies from the IMDb Top 1000 dataset based on their descriptions. To evaluate the accuracy of these predictions, we will use the <code>Llama-3.1-405B-Instruct-Turbo</code> model as a judge, tasked with determining whether the base model’s answers are correct. Since the dataset includes the true genres as ground truth, we can also assess how well the judge model aligns with the actual answers provided in the dataset.</p><p>'''</p>

Comment (joaq-rod) [reply examples/llm-as-a-judge.ipynb:1]: <p>Do you think this could work?</p>

<pre class="ql-syntax" spellcheck="false">def create_tasks(user_contents, task_type, system_prompt, model):
&nbsp; &nbsp; tasks = []
&nbsp; &nbsp; for index, user_content in enumerate(user_contents):
&nbsp; &nbsp; &nbsp; &nbsp; task = {
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; "custom_id": f"{task_type}-{index}",
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; "method": "POST",
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; "url": "/v1/chat/completions",
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; "body": {
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; "model": model,
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; "temperature": 0,
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; "messages": [
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; {"role": "system", "content": system_prompt},
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; {"role": "user", "content": user_content},
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; ],
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; }
&nbsp; &nbsp; &nbsp; &nbsp; }
&nbsp; &nbsp; &nbsp; &nbsp; tasks.append(task)
&nbsp; &nbsp; return tasks


def save_tasks(tasks, task_type):
&nbsp; &nbsp; filename = f"batch_tasks_{task_type}.jsonl"
&nbsp; &nbsp; with open(filename, 'w') as file:
&nbsp; &nbsp; &nbsp; &nbsp; for task in tasks:
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; file.write(json.dumps(task) + '\n')
&nbsp; &nbsp; return filename
</pre><p>And then define prompts as a dict:</p>

<pre class="ql-syntax" spellcheck="false">prompt_dict = {
&nbsp;&nbsp;"ASSISTANT_PROMPT" : '''
&nbsp;&nbsp;&nbsp;&nbsp;You are a helpful assitant that classifies movie genres based on the movie description. Choose one of the following options:&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;Action, Adventure, Animation, Biography, Comedy, Crime, Drama, Family, Fantasy, Film-Noir, History, Horror, Music, Musical, Mystery, Romance, Sci-Fi, Sport, Thriller, War, Western.
&nbsp;&nbsp;&nbsp;&nbsp;Provide your response as a single word with the matching genre. Don't include punctuation.
&nbsp;&nbsp;''',
&nbsp;&nbsp;"USER_CONTENTS" : df['Overview'].tolist()
}

task_list = create_tasks(user_contents=prompt_dict["USER_CONTENTS"], system_prompt=prompt_dict["ASSISTANT_PROMPT"], model="klusterai/Meta-Llama-3.1-8B-Instruct-Turbo", task_type='assistant')
filename = save_tasks(task_list, task_type='assistant')
</pre>

Comment (joaq-rod) [reply examples/llm-as-a-judge.ipynb:1]: Thanks for the idea! Doing that comparison could give the user an idea on a practical implementation of LLM as a judge, but I think adding another model might bring some complications:

1. Consistency with Ground Truth: If we use a larger model for the text classifier, like a 70B one (or tweaking the prompt as i did for the judge model) it might perform so well that the judge has nothing to correct. In that case the judge won’t add much value, it could feel redundant. During the creation of this notebook I already experienced something like that. 

2. Length of teh notebook: The notebook currently has two parts: one for the text classification model and another for the judge model. Adding a third model (or prompt engineering) would increase the number of sections to three, which would also make the notebook longer. It's not something terrible, but it is worth mentioning.

3. Time-to-Market: Adding more complexity means more time to finish the notebook.

The notebook as-is demonstrates how an LLM can be used as a judge to evaluate predictions when there is no ground truth. It provides an example where the predictions align closely with the actual ground truth, showing consistency in the approach (which wasn't that easy to achieve, at least for me hehehe xD).

Rather than offering a complete end-to-end implementation, the goal is to give users a starting point to explore this concept further using kluster API. Users can experiment with improving prompts, trying different models, or applying this method to other use cases. 

Anyway, I'm just giving my opinion here. I'll definitely align with your decision.

Comment (rymc) [reply examples/llm-as-a-judge.ipynb:1]: Good points!
1 .If 8B and 70B are the base models and 405B acts as the judge, the 8B classifier might not perform as well as the 70B (with 405B as the judge). It could be worth investigating this quickly if feasible, as it would demonstrate whether the 70B is the better choice over the 8B in this scenario.
2. Perhaps we could integrate the other model into the existing sections?
3. Agreed—this might work well as a follow-up!

Comment (rymc) [reply examples/llm-as-a-judge.ipynb:1]: <p>Great!</p>

Comment (rymc) [reply examples/llm-as-a-judge.ipynb:1]: <p>yes, much better!</p>

**Review by rymc — APPROVED** [review]
